### PR TITLE
Java ports 1

### PR DESCRIPTION
**Pull request by kwhopper**
Fixes 100's of FileProcessor discrepancies between Java and .NET

### CODE REVIEW

**Comment by kwhopper** [comment]
The first seven commits are bogus hold-overs from getting my fork back to par. Sorry about that

**Comment by drewnoakes** [comment]
> The first seven commits are bogus hold-overs from getting my fork back to par. Sorry about that

It'd be good to clean this up. Can you rebase on master and force push? I've pulled this PR down to my machine as a backup, so don't worry about losing anything. Let me know if I can help.

**Comment by drewnoakes** [comment]
I'm not worried about CodeFactor here. I've tried updating the settings there to not complain about code style. This kind of thing needs to be in the dev's inner loop, not in the CI loop. If we want that, we'll use `StyleCop.Analyzers`.

**Comment by kwhopper** [comment]
The rebase 'seems' reasonable on my local copy of the fork, but it still looks like a mess here. I suspect the rebase was done incorrectly. Unless you have suggestions, maybe I need to re-do my fork entirely?

**Comment by kwhopper** [comment]
Meh. It's a total mess, even locally. Rebasing is still a mystery.

**Comment by kwhopper** [comment]
Too far gone to fix so I'm going to re-do my fork and all these commits. Sorry for the inconvenience.

**Comment by drewnoakes** [comment]
@kwhopper I can't quite deduce what caused the problem here, but this kind of tidyup with git can be tricky. Don't sweat it, I have your commits locally as a backup.

Perhaps you can make a new branch for each logical change and cherry pick commits from the above, then open a PR for each change.

Let me know if you'd like a hand.

File: MetadataExtractor/Formats/Png/PngMetadataReader.cs [hunk 1]
```diff
@@ -182,11 +182,31 @@ private static IEnumerable<Directory> ProcessChunk([NotNull] PngChunk chunk)
                     bytesLeft -= 2;
 
                     var compressedProfile = reader.GetBytes(bytesLeft);
-                    using (var inflaterStream = new DeflateStream(new MemoryStream(compressedProfile), CompressionMode.Decompress))
+                    using (var inflateStream = new DeflateStream(new MemoryStream(compressedProfile), CompressionMode.Decompress))
                     {
-                        var iccDirectory = new IccReader().Extract(new IndexedCapturingReader(inflaterStream));
-                        iccDirectory.Parent = directory;
-                        yield return iccDirectory;
+                        byte[] inflateBuffer = null;
```
[the remark this answers is not in the document]
Comment (kwhopper) [reply MetadataExtractor/Formats/Png/PngMetadataReader.cs]: The issue is you can't yield in the body of a try block with a catch clause. I'll see if there's a different way to accomplish the same thing without a byte array.

File: MetadataExtractor/Formats/Bmp/BmpHeaderDirectory.cs [hunk 2]
```diff
@@ -67,5 +94,119 @@ protected override bool TryGetTagName(int tagType, out string tagName)
...
+            OS2ColorIcon = 0x4943,
+
+            /** "CP" - OS/2 Color pointer */
+            OS2ColorPointer = 0x5043,
+
+            /** "PT" - OS/2 Pointer */
+            OS2Pointer = 0x5450
+        }
+
+        public enum Compression
```
[the remark this answers is not in the document]
Comment (kwhopper) [reply MetadataExtractor/Formats/Bmp/BmpHeaderDirectory.cs]: This was a direct copy from the Java implementation. I'm not sure what the "Bi" means but can remove it if you want.

File: MetadataExtractor/Formats/Bmp/BmpHeaderDirectory.cs [hunk 3]
```diff
@@ -67,5 +94,119 @@ protected override bool TryGetTagName(int tagType, out string tagName)
...
+            /** Super-circle Halftoning */
+            SuperCircle = 3
+        }
+
+        public enum ColorEncoding
+        {
+            Rgb = 0
+        }
+
+        public enum ColorSpaceType : long
```
[the remark this answers is not in the document]
Comment (kwhopper) [reply MetadataExtractor/Formats/Bmp/BmpHeaderDirectory.cs]: BmpHeaderDirectory.TagColorSpaceType is expected to be read as a long in GetColorSpaceTypeDescription, and some of the enum values appear to be outside an int's space (maybe). Either way it seems appropriate to declare this as a long to simplify the switch statement.

File: MetadataExtractor/Formats/Bmp/BmpHeaderDescriptor.cs [hunk 4]
```diff
@@ -71,16 +119,109 @@ public string GetCompressionDescription()
...
+                    return "Unimplemented rendering intent type " + value.ToString();
+            }
+        }
+
+        private static string FormatHex(long value, int digits)
+        {
+            return string.Format("0x{0}", value.ToString("X" + digits.ToString()));
+        }
+
+        public static string FormatFixed1616(long value)
```
[the remark this answers is not in the document]
Comment (kwhopper) [reply MetadataExtractor/Formats/Bmp/BmpHeaderDescriptor.cs]: They're only used once in this class, so I'll remove them as methods and put their code directly inline.

File: MetadataExtractor/Formats/Avi/AviRiffHandler.cs [hunk 5]
```diff
@@ -112,7 +112,7 @@ public void ProcessChunk(string fourCc, byte[] payload)
                             int hours = (int)duration / (int)(Math.Pow(60, 2));
                             int minutes = ((int)duration / (int)(Math.Pow(60, 1))) - (hours * 60);
                             int seconds = (int)Math.Round((duration / (Math.Pow(60, 0))) - (minutes * 60));
-                            string time = new DateTime(2000, 1, 1, hours, minutes, seconds).ToString("hh:mm:ss");
+                            string time = new DateTime(2000, 1, 1, hours, minutes, seconds).ToString("HH:mm:ss");
```
[the remark this answers is not in the document]
Comment (kwhopper) [reply MetadataExtractor/Formats/Avi/AviRiffHandler.cs]: used TimeSpan

File: MetadataExtractor/Formats/Bmp/BmpHeaderDirectory.cs [hunk 6]
```diff
@@ -67,5 +94,119 @@ protected override bool TryGetTagName(int tagType, out string tagName)
         {
             return _tagNameMap.TryGetValue(tagType, out tagName);
         }
+
+        // Possible "magic bytes" indicating bitmap type
```
[the remark this answers is not in the document]
Comment (kwhopper) [reply MetadataExtractor/Formats/Bmp/BmpHeaderDirectory.cs]: done

File: MetadataExtractor/Formats/Bmp/BmpReader.cs [hunk 7]
```diff
@@ -22,105 +22,375 @@
...
+             *    SHORT Height            - Image height in pixels
+             *    WORD  Planes            - Number of color planes
+             *    WORD  BitsPerPixel      - Number of bits per pixel
+             *
+             * OS21XBITMAPHEADER (12 bytes):
+             *
+             *    DWORD  Size             - Size of this structure in bytes
+             *    WORD   Width            - Bitmap width in pixels
+             *    WORD   Height           - Bitmap height in pixel
+             *      WORD   NumPlanes        - Number of bit planes (color depth)
```
[the remark this answers is not in the document]
Comment (kwhopper) [reply MetadataExtractor/Formats/Bmp/BmpReader.cs]: fixed

File: MetadataExtractor/Formats/Bmp/BmpReader.cs [hunk 8]
```diff
@@ -22,105 +22,375 @@
...
+             *    WORD   BitsPerPixel     - Number of bits per pixel per plane
+             *
+             * OS22XBITMAPHEADER (16/64 bytes):
+             *
+             *    DWORD  Size             - Size of this structure in bytes
+             *    DWORD  Width            - Bitmap width in pixels
+             *    DWORD  Height           - Bitmap height in pixel
+             *      WORD   NumPlanes        - Number of bit planes (color depth)
```
[the remark this answers is not in the document]
Comment (kwhopper) [reply MetadataExtractor/Formats/Bmp/BmpReader.cs]: fixed

File: MetadataExtractor/Formats/Exif/makernotes/NikonType2MakernoteDescriptor.cs [hunk 9]
```diff
@@ -98,7 +100,15 @@ public override string GetDescription(int tagType)
...
-            return GetEpochTimeDescription(NikonType2MakernoteDirectory.TagPowerUpTime);
+            // this is generally a byte[] of length 8 directly representing a date and time.
+            // the format is : first 2 bytes together are the year, and then each byte after
+            //                 is month, day, hour, minute, second with the eighth byte unused
+            // e.g., 2011:04:25 01:54:58
+
+            var o = Directory.GetByteArray(NikonType2MakernoteDirectory.TagPowerUpTime);
+            ushort year = ByteConvert.FromBigEndianToNative(BitConverter.ToUInt16(o, 0));
+            return string.Format("{0}:{1}:{2} {3}:{4}:{5}", year, o[2].ToString("D2"), o[3].ToString("D2"),
+                                             o[4].ToString("D2"), o[5].ToString("D2"), o[6].ToString("D2"));
```
[the remark this answers is not in the document]
Comment (kwhopper) [reply MetadataExtractor/Formats/Exif/makernotes/NikonType2MakernoteDescriptor.cs]: fixed, and used interpolation

File: MetadataExtractor/Formats/Riff/RiffReader.cs [hunk 10]
```diff
@@ -70,23 +70,20 @@ public void ProcessRiff([NotNull] SequentialReader reader, [NotNull] IRiffHandle
...
                 if (reader.IsCloserToEnd(8)) return;
 
                 string chunkFourCc = reader.GetString(4, Encoding.ASCII);
                 int chunkSize = reader.GetInt32();
 
-                sizeLeft -= 8;
-
                 // NOTE we fail a negative chunk size here (greater than 0x7FFFFFFF) as we cannot allocate arrays larger than this
-                if (chunkSize < 0 || sizeLeft < chunkSize)
+                if (chunkSize < 0 || sectionSize < chunkSize)
```
[the remark this answers is not in the document]
Comment (kwhopper) [reply MetadataExtractor/Formats/Riff/RiffReader.cs]: Leaving this logic intact.

File: MetadataExtractor/Formats/Riff/RiffReader.cs [hunk 11]
```diff
@@ -96,10 +93,15 @@ public void ProcessChunks([NotNull] SequentialReader reader, int sizeLeft, [NotN
                 {
                     string listName = reader.GetString(4, Encoding.ASCII);
                     if (handler.ShouldAcceptList(listName))
-                        ProcessChunks(reader, sizeLeft - 4, handler);
+                        ProcessChunks(reader, chunkSize - 4, handler);
```
[the remark this answers is not in the document]
Comment (kwhopper) [reply MetadataExtractor/Formats/Riff/RiffReader.cs]: not relevant now

File: MetadataExtractor/Formats/Png/PngMetadataReader.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (drewnoakes) [reply MetadataExtractor/Formats/Png/PngMetadataReader.cs]: I can't remember 100% but I think a newer compiler might be able to do this. We can look at explicitly using a newer version of the compiler via NuGet.

File: MetadataExtractor/Formats/Bmp/BmpHeaderDescriptor.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (drewnoakes) [reply MetadataExtractor/Formats/Bmp/BmpHeaderDescriptor.cs]: Up to you. Having them as methods is nice in that they're self describing. There's no overhead for local functions.

**Comment by drewnoakes** [comment]
I see you've created some new PRs. Thanks! Let's continue the discussion there.

**Comment by drewnoakes** [comment]
I went through your numerous PRs, merged most of them and commented on a few others. Thanks again.

**Pull request closed without merging**